views/default/page/layouts/widgets/add_panel.php

```diff
@@ -1,69 +1,87 @@
 <?php
 /**
  * Widget add panel
- *
- * @uses $vars['widgets']     Array of current widgets
- * @uses $vars['context']     The context for this widget layout
- * @uses $vars['exact_match'] Only use widgets that match the context
- * @uses $vars['container']   Container to optional limit widgets for. Defaults to page_owner_entity
  */
 
-$widgets = elgg_extract('widgets', $vars);
+if (!elgg_is_xhr()) {
+	echo elgg_echo('ajax:not_is_xhr');
+	return;
+}
+
 $context = elgg_extract('context', $vars);
-$exact = elgg_extract('exact_match', $vars, false);
-$container = elgg_extract('container', $vars, elgg_get_page_owner_entity());
+$exact = elgg_extract('exact_match', $vars);
+$owner_guid = (int) elgg_extract('owner_guid', $vars);
+$owner = get_entity($owner_guid);
+elgg_set_page_owner_guid($owner_guid);
+
+$context_stack = elgg_extract('context_stack', $vars);
+if (!empty($context_stack)) {
+	elgg_set_context_stack($context_stack);
+}
 
+$widgets = elgg_get_widgets($owner->guid, $context);
 $widget_types = elgg_get_widget_types([
 	'context' => $context,
 	'exact' => $exact,
-	'container' => $container,
+	'container' => $owner,
 ]);
 uasort($widget_types, function($a, $b) {
-	return strcmp($a->name,$b->name);
+	return strcmp($a->name, $b->name);
 });
 
-$current_handlers = array();
+$current_handlers = [];
 foreach ($widgets as $column_widgets) {
 	foreach ($column_widgets as $widget) {
 		$current_handlers[] = $widget->handler;
 	}
 }
 
-?>
-<div class="elgg-widgets-add-panel hidden clearfix" id="widgets-add-panel">
-	<p>
-		<?php echo elgg_echo('widgets:add:description'); ?>
-	</p>
-	<ul>
-<?php
-		foreach ($widget_types as $handler => $widget_type) {
-			// check if widget added and only one instance allowed
-			if ($widget_type->multiple == false && in_array($handler, $current_handlers)) {
-				$class = 'elgg-state-unavailable';
-				$tooltip = elgg_echo('widget:unavailable');
-			} else {
-				$class = 'elgg-state-available';
-				$tooltip = $widget_type->description;
-			}
+$result = elgg_autop(elgg_echo('widgets:add:description'));
+
+$list_items = '';
+foreach ($widget_types as $handler => $widget_type) {
+	$class = [];
+	// check if widget added and only one instance allowed
+	if (!$widget_type->multiple && in_array($handler, $current_handlers)) {
+		$class[] = 'elgg-state-unavailable';
+	} else {
+		$class[] = 'elgg-state-available';
+	}
+
+	$class[] = $widget_type->multiple ? 'elgg-widget-multiple' : 'elgg-widget-single';
+	
+	$item_content = '<div class="elgg-widgets-add-actions float-alt">';
```

Comment:
Make it into a menu and float it with CSS if needed. I just don't want any more inline helper class uses. Makes it impossible for other plugin to work with without replacing the view.